compiler/rustc_trait_selection/src/opaque_types.rs

```diff
@@ -730,15 +734,22 @@ where
                 // Skip lifetime parameters of the enclosing item(s)
                 // Also skip the witness type, because that has no free regions.
 
-                substs.as_generator().tupled_upvars_ty().visit_with(self);
+                // We need to shallow resolve to ensure tuple_upvars_ty is of kind Tuple
+                // At this point, the kind should not be of type Infer but could potentially
+                // be of kind Error
+                let ty = self.infcx.shallow_resolve(substs.as_generator().tupled_upvars_ty());
 
-                for upvar_ty in substs.as_generator().upvar_tys() {
-                    upvar_ty.visit_with(self);
-                }
+                if let ty::Tuple(_) = ty.kind() {
```

Comment:
It would be better to write the code more robustly, I think, so that it permits `Error` or `Tuple` but not other cases. I suspect the right fix is going to be to modify `upvar_tys` -- wouldn't we expect ICEs from other callers?